component/loki/source/docker/docker.go

```diff
@@ -0,0 +1,261 @@
+	"github.com/grafana/agent/component/discovery"
+	dt "github.com/grafana/agent/component/loki/source/docker/internal/dockertarget"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/relabel"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "loki.source.docker",
+		Args: Arguments{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+const (
+	dockerLabel                = model.MetaLabelPrefix + "docker_"
+	dockerLabelContainerPrefix = dockerLabel + "container_"
+	dockerLabelContainerID     = dockerLabelContainerPrefix + "id"
+)
+
+// Arguments holds values which are used to configure the loki.source.docker
+// component.
+type Arguments struct {
+	Host         string              `river:"host,attr"`
+	Targets      []discovery.Target  `river:"targets,attr"`
+	ForwardTo    []loki.LogsReceiver `river:"forward_to,attr"`
+	Labels       map[string]string   `river:"labels,attr,optional"`
+	RelabelRules flow_relabel.Rules  `river:"relabel_rules,attr,optional"`
+}
+
+var (
+	_ component.Component      = (*Component)(nil)
+	_ component.DebugComponent = (*Component)(nil)
+)
+
+// Component implements the loki.source.file component.
+type Component struct {
+	opts    component.Options
+	metrics *dt.Metrics
+
+	mut           sync.RWMutex
+	args          Arguments
+	manager       *Manager
+	lastOptions   *Options
+	handler       loki.LogsReceiver
+	receivers     []loki.LogsReceiver
+	posFile       positions.Positions
+	rcs           []*relabel.Config
+	defaultLabels model.LabelSet
+}
+
+// New creates a new loki.source.file component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	err := os.MkdirAll(o.DataPath, 0750)
+	if err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+	positionsFile, err := positions.New(o.Logger, positions.Config{
+		SyncPeriod:        10 * time.Second,
+		PositionsFile:     filepath.Join(o.DataPath, "positions.yml"),
+		IgnoreInvalidYaml: false,
+		ReadOnly:          false,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Component{
+		opts:    o,
+		metrics: dt.NewMetrics(o.Registerer),
+
+		handler:   make(loki.LogsReceiver),
+		receivers: args.ForwardTo,
+		posFile:   positionsFile,
+	}
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	defer c.posFile.Stop()
+
+	defer func() {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+
+		// Guard for safety, but it's not possible for Run to be called without
+		// c.tailer being initialized.
+		if c.manager != nil {
+			c.manager.Stop()
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			for _, receiver := range c.receivers {
```

Comment:
Potential race condition here with accessing `c.receivers`; you should grab the mutex when reading:

```go 
c.mut.RLock()
receivers := c.receivers 
c.mut.RUnlock()

for _, receiver := range receivers {
   ... 
}
```